include/dxc/Support/dxcapi.use.h

```diff
@@ -19,9 +19,51 @@ namespace dxc {
 extern const char *kDxCompilerLib;
 extern const char *kDxilLib;
 
-// Helper class to dynamically load the dxcompiler or a compatible libraries.
-class DxcDllSupport {
+// Interface for common dll operations
+class DllLoader {
+public:
+  virtual HRESULT OverrideDll(LPCSTR dll, LPCSTR entryPoint) = 0;
+  DllLoader() = default;
+  DllLoader(const DllLoader &) = delete;
+  DllLoader(DllLoader &&) = delete;
+
 protected:
+  virtual HRESULT CreateInstanceImpl(REFCLSID clsid, REFIID riid,
+                                     IUnknown **pResult) = 0;
+  virtual HRESULT CreateInstance2Impl(IMalloc *pMalloc, REFCLSID clsid,
+                                      REFIID riid, IUnknown **pResult) = 0;
+
+public:
+  template <typename TInterface>
+  HRESULT CreateInstance(REFCLSID clsid, TInterface **pResult) {
+    return CreateInstanceImpl(clsid, __uuidof(TInterface),
+                              (IUnknown **)pResult);
+  }
+  HRESULT CreateInstance(REFCLSID clsid, REFIID riid, IUnknown **pResult) {
+    return CreateInstanceImpl(clsid, riid, (IUnknown **)pResult);
+  }
+
+  template <typename TInterface>
+  HRESULT CreateInstance2(IMalloc *pMalloc, REFCLSID clsid,
+                          TInterface **pResult) {
+    return CreateInstance2Impl(pMalloc, clsid, __uuidof(TInterface),
+                               (IUnknown **)pResult);
+  }
+  HRESULT CreateInstance2(IMalloc *pMalloc, REFCLSID clsid, REFIID riid,
+                          IUnknown **pResult) {
+    return CreateInstance2Impl(pMalloc, clsid, riid, (IUnknown **)pResult);
+  }
+
+  virtual bool HasCreateWithMalloc() const = 0;
+
+  virtual bool IsEnabled() const = 0;
+
+  virtual HMODULE Detach() = 0;
```

Comment:
I think that Detach can also be removed from the interface, this is only ever called directly on concrete `SpecificDllLoader` instances.